MySQL connector 3.6 cannot parse DDL from a server running with `sql_mode=ANSI_QUOTES`

# MySQL connector 3.6 cannot parse DDL from a server running with `sql_mode=ANSI_QUOTES`

## Bug report

**What Debezium connector do you use and what version?**

MySQL connector (`io.debezium.connector.mysql.MySqlConnector`), **3.6.0.Final**
(`quay.io/debezium/connect:3.6.0.Final`, digest `sha256:6f3fe640…0943d47`), and also
**3.7.0-SNAPSHOT** built from `main` at commit `a0fd78a46c`.

Works in 3.5.0.Final, so this seems to be a regression in 3.6.

AI disclosure: used Claude Opus 5 to create and run through the variants of the reproducer. 

---

**What is the connector configuration?**

The unmodified tutorial configuration
([`register-mysql.json`](https://github.com/debezium/debezium-examples/blob/main/tutorial/register-mysql.json)):

```json
{
    "name": "inventory-connector",
    "config": {
        "connector.class": "io.debezium.connector.mysql.MySqlConnector",
        "tasks.max": "1",
        "database.hostname": "mysql",
        "database.port": "3306",
        "database.user": "debezium",
        "database.password": "dbz",
        "database.server.id": "184054",
        "topic.prefix": "dbserver1",
        "database.include.list": "inventory",
        "schema.history.internal.kafka.bootstrap.servers": "kafka:9092",
        "schema.history.internal.kafka.topic": "schema-changes.inventory"
    }
}
```

Adding `"ddl.parser.type": "legacy"` to this exact configuration makes the problem
go away (see below).

---

**What is the captured database version and mode of deployment?**

`quay.io/debezium/example-mysql:3.6.0.Final` (MySQL **8.2.0**, `binlog_format=ROW`),
run locally in Docker, i.e. the tutorial deployment.

The only deviation from the tutorial is the server's `sql_mode`:

```
ANSI_QUOTES,ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION
```

i.e. MySQL's defaults plus `ANSI_QUOTES`.

---

**What behavior do you expect?**

The connector snapshots and streams `inventory` normally, exactly as it does against a
server with the default `sql_mode`. Under `ANSI_QUOTES`, MySQL treats `"` as an
identifier quote, so `SHOW CREATE TABLE` and the DDL recorded in the binlog use `"…"`
instead of `` `…` `` for identifiers. The DDL parser should accept that quoting style.

---

**What behavior do you see?**

The task fails immediately during the initial snapshot, while reading the table
structure. The parser is fed a **corrupted** statement: every double-quoted
*identifier* has been rewritten into a single-quoted *string literal*, which is not
valid SQL anywhere:

```
io.debezium.text.ParsingException: DDL statement couldn't be parsed. Please open a GitHub issue
at https://github.com/debezium/dbz/issues with the statement 'CREATE TABLE 'addresses' (
  'id' int NOT NULL AUTO_INCREMENT,
  'customer_id' int NOT NULL,
  'street' varchar(255) NOT NULL,
  'city' varchar(255) NOT NULL,
  'state' varchar(255) NOT NULL,
  'zip' varchar(255) NOT NULL,
  'type' enum('SHIPPING','BILLING','LIVING') NOT NULL,
  PRIMARY KEY ('id'),
  KEY 'address_customer' ('customer_id'),
  CONSTRAINT 'addresses_ibfk_1' FOREIGN KEY ('customer_id') REFERENCES 'customers' ('id')
) ENGINE=InnoDB AUTO_INCREMENT=17 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci'
mismatched input ''addresses'' expecting {'.', ACCOUNT_SYMBOL, …, IDENTIFIER, BACK_TICK_QUOTED_ID, DOUBLE_QUOTED_TEXT}
```

What MySQL actually returned was well-formed:

```sql
mysql> SHOW CREATE TABLE inventory.customers\G
Create Table: CREATE TABLE "customers" (
  "id" int NOT NULL AUTO_INCREMENT,
  "first_name" varchar(255) NOT NULL,
  ...
  PRIMARY KEY ("id"),
  UNIQUE KEY "email" ("email")
) ENGINE=InnoDB ...
```

Note the last token in the ANTLR "expecting" set: `DOUBLE_QUOTED_TEXT`. The grammar
*would* have accepted `"addresses"` as an identifier — the statement is broken before
it reaches the parser.

**Streaming is affected too**, not just `SHOW CREATE TABLE`. On a server with the
default `sql_mode` (so the snapshot succeeds), a client that uses `ANSI_QUOTES` puts
double-quoted identifiers into the binlog, and the connector then dies mid-stream:

```sql
SET SESSION sql_mode='ANSI_QUOTES';
ALTER TABLE inventory."customers" ADD COLUMN "middle_name" varchar(255) NULL;
```

```
io.debezium.text.ParsingException: DDL statement couldn't be parsed. … with the statement
'ALTER TABLE inventory.'customers' ADD COLUMN 'middle_name' varchar(255) NULL'
extraneous input '.' expecting {<EOF>, '(', ALTER_SYMBOL, …}
```

So there is no server-side or session-side workaround: any `ANSI_QUOTES` DDL that
reaches the connector seems to be mangled.

**Suspected cause.** The 3.6 migration from the Positive Technologies grammar to the
Oracle MySQL grammar (`debezium/dbz#1216`) instantiates the
lexer/parser permanently in `ANSI_QUOTES` mode
(`MySqlLexerBase`/`MySqlParserBase`: `SqlModes.sqlModeFromString("ANSI_QUOTES")`) and
compensates for MySQL's *default* mode by rewriting the DDL text before parsing —
`MySqlAntlrDdlParser.parse()` calls `DdlNormalizer.normalize()`, which turns every
`"…"` run into `'…'`. That rewrite is unconditional, so it corrupts input from servers
where `"…"` really is an identifier. The connector knows the server's `sql_mode` (it
reads it into `BinlogSystemVariables`), but the normalizer does not consult it.

---

**Do you see the same behaviour using the latest released Debezium version?**

Yes — reproduced both on 3.6.0.Final and on a **`main` build (3.7.0-SNAPSHOT)** at
commit `a0fd78a46c` (2026-07-24), which already contains the recent `DdlNormalizer`
fixes [#7682](https://github.com/debezium/debezium/pull/7682) (`dbz#2237`, merged
2026-07-20) and [#7701](https://github.com/debezium/debezium/pull/7701) (`dbz#2254`,
merged 2026-07-23). Those fix other effects of the same rewrite (`""` inside
single-quoted strings; reserved-keyword quoting leaking into comments and literals) but
do not touch the `ANSI_QUOTES` case — the rewrite is still unconditional.

Results of the same reproducer, same MySQL server (`sql_mode` including `ANSI_QUOTES`),
same connector config:

| Connect version | `ddl.parser.type` | Result |
| --- | --- | --- |
| **3.7.0-SNAPSHOT** (`main`, `a0fd78a46c`) | `default` (Oracle grammar) | **task FAILED** — same `ParsingException` |
| 3.7.0-SNAPSHOT (`main`, `a0fd78a46c`) | `legacy` (PT grammar) | task RUNNING, snapshot OK |
| 3.6.0.Final | `default` (Oracle grammar) | **task FAILED** — `ParsingException` during snapshot |
| 3.6.0.Final | `legacy` (PT grammar) | task RUNNING, snapshot + streaming OK |
| 3.5.0.Final | n/a | task RUNNING, snapshot + streaming OK |

Control: 3.6.0.Final with `ddl.parser.type=default` against the *same* server with
`ANSI_QUOTES` removed from `sql_mode` runs fine — `ANSI_QUOTES` is the only variable.

`main`'s `DdlNormalizer` can also be shown to mangle the input in isolation — it has no
dependencies beyond `java.util.regex`, so it compiles and runs standalone:

```
input:  CREATE TABLE "customers" ("id" int NOT NULL AUTO_INCREMENT, PRIMARY KEY ("id"))
output: CREATE TABLE 'customers' ('id' int NOT NULL AUTO_INCREMENT, PRIMARY KEY ('id'))

input:  ALTER TABLE inventory."customers" ADD COLUMN "middle_name" varchar(255) NULL
output: ALTER TABLE inventory.'customers' ADD COLUMN 'middle_name' varchar(255) NULL
```

Both the `legacy` parser on 3.6.0.Final and 3.5.0.Final also handled `ANSI_QUOTES`
DDL arriving via the binlog correctly, including the schema change:

```
op c after {"id": 1006, "first_name": "Bo", "last_name": "Quote",
            "email": "bo@example.com", "middle_name": null, "nickname": "bobo"}
```

after `SET SESSION sql_mode='ANSI_QUOTES'; ALTER TABLE inventory."customers" ADD COLUMN "nickname" varchar(50) NULL;`
followed by an insert — i.e. both `ANSI_QUOTES` `ALTER`s were applied to the schema
correctly by those two, and both `middle_name` and `nickname` appear in the event.

---

**Do you have the connector logs, ideally from start till finish?**

Relevant excerpt (full logs can be attached on request):

```
2026-07-25T15:07:26,862 INFO   MySQL||snapshot  All eligible tables schema should be captured, capturing: [inventory.addresses, inventory.customers, inventory.geom, inventory.orders, inventory.products, inventory.products_on_hand]   [io.debezium.connector.binlog.BinlogSnapshotChangeEventSource]
2026-07-25T15:07:27,187 INFO   MySQL||snapshot  Reading structure of database 'inventory'   [io.debezium.connector.binlog.BinlogSnapshotChangeEventSource]
2026-07-25T15:07:27,199 ERROR  MySQL||snapshot  Error during snapshot   [io.debezium.relational.RelationalSnapshotChangeEventSource]
io.debezium.text.ParsingException: DDL statement couldn't be parsed. … with the statement 'CREATE TABLE 'addresses' ( …
        at io.debezium.antlr.ParsingErrorListener.syntaxError(ParsingErrorListener.java:44)
        at io.debezium.ddl.parser.mysql.generated.MySqlParser.createTable(MySqlParser.java:6271)
        at io.debezium.connector.mysql.antlr.MySqlAntlrDdlParser.parseTree(MySqlAntlrDdlParser.java:101)
        at io.debezium.antlr.AntlrDdlParser.parse(AntlrDdlParser.java:74)
        at io.debezium.connector.mysql.antlr.MySqlAntlrDdlParser.parse(MySqlAntlrDdlParser.java:131)
        at io.debezium.connector.binlog.BinlogDatabaseSchema.parseDdl(BinlogDatabaseSchema.java:316)
        at io.debezium.connector.binlog.BinlogDatabaseSchema.parseSnapshotDdl(BinlogDatabaseSchema.java:248)
        at io.debezium.connector.binlog.BinlogSnapshotChangeEventSource.addSchemaEvent(BinlogSnapshotChangeEventSource.java:282)
        at io.debezium.connector.binlog.BinlogSnapshotChangeEventSource.readTableStructure(BinlogSnapshotChangeEventSource.java:355)
        at io.debezium.relational.RelationalSnapshotChangeEventSource.doExecute(RelationalSnapshotChangeEventSource.java:182)
        ...
Caused by: org.antlr.v4.runtime.InputMismatchException
2026-07-25T15:07:27,270 ERROR  MySQL||lifecycle  WorkerSourceTask{id=inventory-connector-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted   [org.apache.kafka.connect.runtime.WorkerTask]
```

---

**How to reproduce the issue using our [tutorial](https://github.com/debezium/debezium-examples/tree/main/tutorial) deployment?**

Start from `debezium-examples/tutorial` with `DEBEZIUM_VERSION=3.6.0.Final`.

```bash
cd debezium-examples/tutorial
export DEBEZIUM_VERSION=3.6.0.Final

# 1. Bring up the tutorial as documented. MySQL must start with its default
#    sql_mode here, because the image's inventory.sql seed script uses
#    double-quoted string literals and would fail under ANSI_QUOTES.
docker compose -f docker-compose-mysql.yaml up -d
# wait until the seed data is there
until docker compose -f docker-compose-mysql.yaml exec -T mysql \
    mysql -uroot -pdebezium -e 'select count(*) from inventory.customers'; do sleep 2; done

# 2. Switch the server to ANSI_QUOTES and restart it, so MySQL *starts* in that mode.
docker compose -f docker-compose-mysql.yaml exec -T mysql mysql -uroot -pdebezium -e \
  "SET PERSIST sql_mode='ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION,ANSI_QUOTES';"
docker compose -f docker-compose-mysql.yaml restart mysql

# confirm the server now quotes identifiers with "
docker compose -f docker-compose-mysql.yaml exec -T mysql mysql -uroot -pdebezium -e \
  'select @@global.sql_mode\G show create table inventory.customers\G'

# 3. Register the unmodified tutorial connector.
curl -i -X POST -H "Accept:application/json" -H "Content-Type:application/json" \
  localhost:8083/connectors/ -d @register-mysql.json

# 4. The task is FAILED with the ParsingException above.
curl -s localhost:8083/connectors/inventory-connector/status | jq '.tasks[0] | {state, trace}'
```

**Confirming the parser is the trigger.** Register the same configuration with
`"ddl.parser.type": "legacy"` added (and a distinct `name`, `topic.prefix`,
`database.server.id` and `schema.history.internal.kafka.topic` so the two connectors do
not collide) — that task reaches `RUNNING` and snapshots all six `inventory` tables:

```bash
curl -s -X POST -H "Content-Type:application/json" localhost:8083/connectors/ \
  -d @register-mysql-legacy.json
curl -s localhost:8083/connectors/inventory-connector-legacy/status | jq '.tasks[0].state'
# "RUNNING"
docker compose -f docker-compose-mysql.yaml exec -T kafka /kafka/bin/kafka-console-consumer.sh \
  --bootstrap-server kafka:9092 --topic dbzlegacy.inventory.customers \
  --from-beginning --max-messages 1
# op r after {"id": 1001, "first_name": "Sally", ...}
```

**Comparing against 3.5.0.Final.** Add a second worker to the same compose project, so
it captures the identical server:

```yaml
# docker-compose-connect35.yaml
services:
  connect35:
    image: quay.io/debezium/connect:3.5.0.Final
    ports:
     - 8084:8083
    environment:
     - BOOTSTRAP_SERVERS=kafka:9092
     - GROUP_ID=35
     - CONFIG_STORAGE_TOPIC=connect35_configs
     - OFFSET_STORAGE_TOPIC=connect35_offsets
     - STATUS_STORAGE_TOPIC=connect35_statuses
```

```bash
docker compose -f docker-compose-mysql.yaml -f docker-compose-connect35.yaml up -d connect35
# register the tutorial config (distinct name/topic.prefix/server.id/history topic) on :8084
curl -s localhost:8084/connectors/inventory-connector-35/status | jq '.tasks[0] | {version, state}'
# { "version": "3.5.0.Final", "state": "RUNNING" }
```

**Running the same reproducer against `main`.** Build just the MySQL connector plugin
per [CONTRIBUTING.md](https://github.com/debezium/debezium/blob/main/CONTRIBUTING.md)
and mount the assembled plugin directory over the one in the worker image:

```bash
# in a debezium/debezium checkout of main
./mvnw clean install -Passembly -DskipITs -DskipTests -pl debezium-connector-mysql -am
mkdir -p debezium-connector-mysql/target/plugin
tar xzf debezium-connector-mysql/target/debezium-connector-mysql-*-plugin.tar.gz \
    -C debezium-connector-mysql/target/plugin
```

```yaml
# docker-compose-connect-main.yaml
services:
  connect-main:
    image: quay.io/debezium/connect:3.6.0.Final
    ports:
     - 8085:8083
    volumes:
     - ${DEBEZIUM_MAIN_PLUGIN}:/kafka/connect/debezium-connector-mysql:ro
    environment:
     - BOOTSTRAP_SERVERS=kafka:9092
     - GROUP_ID=main
     - CONFIG_STORAGE_TOPIC=connectmain_configs
     - OFFSET_STORAGE_TOPIC=connectmain_offsets
     - STATUS_STORAGE_TOPIC=connectmain_statuses
```

```bash
export DEBEZIUM_MAIN_PLUGIN=<debezium checkout>/debezium-connector-mysql/target/plugin/debezium-connector-mysql
docker compose -f docker-compose-mysql.yaml -f docker-compose-connect-main.yaml up -d connect-main
curl -s localhost:8085/connector-plugins | jq -r '.[] | select(.class|endswith("MySqlConnector")) | "\(.class) \(.version)"'
# io.debezium.connector.mysql.MySqlConnector 3.7.0-SNAPSHOT
# then register the tutorial config (distinct name/topic.prefix/server.id/history topic) on :8085
```

The whole build takes under a minute, and the plugin tarball is self-contained, so no
worker image rebuild is needed.

**Streaming variant.** Put `sql_mode` back to the default and restart `mysql`, so a
`default`-parser connector snapshots successfully (this is also the control run — it
reaches `RUNNING`). Then have a *client* use `ANSI_QUOTES`:

```bash
docker compose -f docker-compose-mysql.yaml exec -T mysql mysql -uroot -pdebezium -e \
 'SET SESSION sql_mode="ANSI_QUOTES";
  ALTER TABLE inventory."customers" ADD COLUMN "middle_name" varchar(255) NULL;
  INSERT INTO inventory."customers" ("first_name","last_name","email","middle_name")
    VALUES ("Ann","Ansi","ann@example.com","Q");'
```

The `default`-parser task fails on the next binlog DDL event with
`ALTER TABLE inventory.'customers' ADD COLUMN 'middle_name' varchar(255) NULL`, while
the `legacy` and 3.5.0.Final connectors keep streaming and pick up the new column.

--- a/debezium-connector-binlog/src/main/java/io/debezium/connector/binlog/BinlogSnapshotChangeEventSource.java
+++ b/debezium-connector-binlog/src/main/java/io/debezium/connector/binlog/BinlogSnapshotChangeEventSource.java
@@ -41,6 +41,7 @@ import io.debezium.DebeziumException;
 import io.debezium.connector.SnapshotRecord;
 import io.debezium.connector.binlog.jdbc.BinlogConnectorConnection;
 import io.debezium.connector.binlog.jdbc.BinlogConnectorConnection.DatabaseLocales;
+import io.debezium.connector.binlog.jdbc.BinlogSystemVariables;
 import io.debezium.connector.binlog.metrics.BinlogSnapshotChangeEventSourceMetrics;
 import io.debezium.data.Envelope;
 import io.debezium.function.BlockingConsumer;
@@ -315,6 +316,8 @@ public abstract class BinlogSnapshotChangeEventSource<P extends BinlogPartition,
         if (!snapshottingTask.isOnDemand()) {
             // Record default charset
             addSchemaEvent(snapshotContext, "", connection.setStatementFor(connection.readCharsetSystemVariables()));
+            // Set sql_mode directly so DDL parser knows whether ANSI_QUOTES is active
+            databaseSchema.setSystemVariables(BinlogSystemVariables.BinlogScope.GLOBAL, connection.readSqlModeSystemVariable());
         }
 
         for (TableId tableId : capturedSchemaTables) {

--- a/debezium-connector-binlog/src/main/java/io/debezium/connector/binlog/jdbc/BinlogConnectorConnection.java
+++ b/debezium-connector-binlog/src/main/java/io/debezium/connector/binlog/jdbc/BinlogConnectorConnection.java
@@ -47,6 +47,7 @@ public abstract class BinlogConnectorConnection extends JdbcConnection {
 
     private static final String SQL_SHOW_SYSTEM_VARIABLES = "SHOW VARIABLES";
     private static final String SQL_SHOW_SYSTEM_VARIABLES_CHARACTER_SET = "SHOW VARIABLES WHERE Variable_name IN ('character_set_server','collation_server')";
+    private static final String SQL_SHOW_SYSTEM_VARIABLES_SQL_MODE = "SHOW VARIABLES WHERE Variable_name = 'sql_mode'";
     private static final String SQL_SHOW_SESSION_VARIABLE_SSL_VERSION = "SHOW SESSION STATUS LIKE 'Ssl_version'";
     private static final String QUOTED_CHARACTER = "`";
     public static final String MASTER_STATUS_STATEMENT = "SHOW MASTER STATUS";
@@ -246,6 +247,16 @@ public abstract class BinlogConnectorConnection extends JdbcConnection {
         // Read the system variables from the MySQL instance and get the current database name ...
         LOGGER.debug("Reading charset-related system variables before parsing DDL history.");
         return querySystemVariables(SQL_SHOW_SYSTEM_VARIABLES_CHARACTER_SET);
+    }
+
+    /**
+     * Read the sql_mode system variable from the server.
+     *
+     * @return a map containing the sql_mode variable; never null
+     */
+    public Map<String, String> readSqlModeSystemVariable() {
+        LOGGER.debug("Reading sql_mode system variable before parsing DDL history.");
+        return querySystemVariables(SQL_SHOW_SYSTEM_VARIABLES_SQL_MODE);
     }
 
     /**

--- a/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/antlr/DdlNormalizer.java
+++ b/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/antlr/DdlNormalizer.java
@@ -14,6 +14,10 @@ import java.util.regex.Pattern;
  * but MySQL's default mode allows double quotes for strings. This normalizer
  * transforms DDL before parsing to support both modes.
  *
+ * When the server is running with {@code sql_mode=ANSI_QUOTES}, double-quoted
+ * text represents identifiers and must not be converted. In that case, callers
+ * should pass {@code ansiQuotesMode=true} to skip the conversion.
+ *
  * Also adds backticks around reserved keywords when used as identifiers.
  *
  * @author Debezium Authors
@@ -48,13 +52,28 @@ public class DdlNormalizer {
     /**
      * Normalizes MySQL DDL by converting double-quoted strings to single-quoted strings
      * while preserving backtick-quoted identifiers, and adding backticks around reserved
-     * keywords when used as identifiers.
+     * keywords when used as identifiers. Assumes the server is NOT in ANSI_QUOTES mode.
      *
      * @param ddlContent The DDL statement to normalize
      * @return Normalized DDL with double-quoted strings converted to single quotes
      *         and reserved keywords backtick-quoted, or the original input if null or empty
      */
     public static String normalize(String ddlContent) {
+        return normalize(ddlContent, false);
+    }
+
+    /**
+     * Normalizes MySQL DDL by adding backticks around reserved keywords when used as
+     * identifiers. When {@code ansiQuotesMode} is false, also converts double-quoted
+     * strings to single-quoted strings (for servers using the default sql_mode where
+     * double quotes delimit string literals). When {@code ansiQuotesMode} is true,
+     * double-quoted text is left as-is because it represents identifiers.
+     *
+     * @param ddlContent The DDL statement to normalize
+     * @param ansiQuotesMode true if the server's sql_mode includes ANSI_QUOTES
+     * @return Normalized DDL, or the original input if null or empty
+     */
+    public static String normalize(String ddlContent, boolean ansiQuotesMode) {
         if (ddlContent == null || ddlContent.isEmpty()) {
             return ddlContent;
         }
@@ -66,6 +85,9 @@ public class DdlNormalizer {
         while (matcher.find()) {
             normalized.append(addBackticksToReservedKeywords(ddlContent.substring(lastEnd, matcher.start())));
             if (isComment(matcher) || isSingleQuotedString(matcher) || isBacktickIdentifier(matcher)) {
+                preserveOriginal(matcher, normalized);
+            }
+            else if (ansiQuotesMode) {
                 preserveOriginal(matcher, normalized);
             }
             else {

--- a/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/antlr/MySqlAntlrDdlParser.java
+++ b/debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/antlr/MySqlAntlrDdlParser.java
@@ -10,6 +10,7 @@ import java.sql.Types;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Objects;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.stream.Collectors;
@@ -25,6 +26,8 @@ import io.debezium.antlr.AntlrDdlParser;
 import io.debezium.antlr.AntlrDdlParserListener;
 import io.debezium.antlr.DataTypeResolver;
 import io.debezium.antlr.DataTypeResolver.DataTypeEntry;
+import io.debezium.antlr.mysql.SqlMode;
+import io.debezium.antlr.mysql.SqlModes;
 import io.debezium.connector.binlog.charset.BinlogCharsetRegistry;
 import io.debezium.connector.binlog.jdbc.BinlogSystemVariables;
 import io.debezium.connector.mysql.antlr.listener.MySqlAntlrDdlParserListener;
@@ -126,9 +129,17 @@ public class MySqlAntlrDdlParser extends AntlrDdlParser<MySqlLexer, MySqlParser>
 
     @Override
     public DdlChanges parse(String ddlContent, Tables databaseTables) {
-        // Normalize double-quoted strings to support MySQL default mode in ANSI_QUOTES parser
-        String normalizedDdl = DdlNormalizer.normalize(ddlContent);
+        String normalizedDdl = DdlNormalizer.normalize(ddlContent, isAnsiQuotesMode());
         return super.parse(normalizedDdl, databaseTables);
+    }
+
+    private boolean isAnsiQuotesMode() {
+        String sqlMode = systemVariables.getVariable("sql_mode");
+        if (sqlMode == null || sqlMode.isEmpty()) {
+            return false;
+        }
+        Set<SqlMode> modes = SqlModes.sqlModeFromString(sqlMode);
+        return modes.contains(SqlMode.AnsiQuotes);
     }
 
     @Override

--- a/debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlAntlrDdlParserTest.java
+++ b/debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlAntlrDdlParserTest.java
@@ -247,4 +247,102 @@ public class MySqlAntlrDdlParserTest
             this.ddlChanges = changesListener;
         }
     }
+
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void shouldParseCreateTableWithAnsiQuotesMode() {
+        parser.parse("SET sql_mode='ANSI_QUOTES'", tables);
+        String ddl = "CREATE TABLE \"customers\" (\n" +
+                "  \"id\" int NOT NULL AUTO_INCREMENT,\n" +
+                "  \"first_name\" varchar(255) NOT NULL,\n" +
+                "  \"last_name\" varchar(255) NOT NULL,\n" +
+                "  \"email\" varchar(255) NOT NULL,\n" +
+                "  PRIMARY KEY (\"id\"),\n" +
+                "  UNIQUE KEY \"email\" (\"email\")\n" +
+                ") ENGINE=InnoDB AUTO_INCREMENT=1005 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci";
+        parser.parse(ddl, tables);
+        assertThat(parser.getParsingExceptionsFromWalker()).isEmpty();
+
+        Table table = tables.forTable(null, null, "customers");
+        assertThat(table).isNotNull();
+        assertThat(table.columns()).hasSize(4);
+        assertThat(table.columnWithName("id")).isNotNull();
+        assertThat(table.columnWithName("first_name")).isNotNull();
+        assertThat(table.columnWithName("last_name")).isNotNull();
+        assertThat(table.columnWithName("email")).isNotNull();
+        assertThat(table.primaryKeyColumnNames()).containsExactly("id");
+    }
+
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void shouldParseAlterTableWithAnsiQuotesMode() {
+        parser.parse("SET sql_mode='ANSI_QUOTES'", tables);
+        parser.parse("CREATE TABLE \"customers\" (\n" +
+                "  \"id\" int NOT NULL AUTO_INCREMENT,\n" +
+                "  \"first_name\" varchar(255) NOT NULL,\n" +
+                "  PRIMARY KEY (\"id\")\n" +
+                ") ENGINE=InnoDB", tables);
+        parser.parse("ALTER TABLE \"customers\" ADD COLUMN \"middle_name\" varchar(255) NULL", tables);
+        assertThat(parser.getParsingExceptionsFromWalker()).isEmpty();
+
+        Table table = tables.forTable(null, null, "customers");
+        assertThat(table).isNotNull();
+        assertThat(table.columns()).hasSize(3);
+        assertThat(table.columnWithName("middle_name")).isNotNull();
+        assertThat(table.columnWithName("middle_name").isOptional()).isTrue();
+    }
+
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void shouldParseCreateTableWithAnsiQuotesModeAndFullSqlMode() {
+        parser.parse("SET sql_mode='ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE," +
+                "ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION,ANSI_QUOTES'", tables);
+        String ddl = "CREATE TABLE \"addresses\" (\n" +
+                "  \"id\" int NOT NULL AUTO_INCREMENT,\n" +
+                "  \"customer_id\" int NOT NULL,\n" +
+                "  \"street\" varchar(255) NOT NULL,\n" +
+                "  \"city\" varchar(255) NOT NULL,\n" +
+                "  \"state\" varchar(255) NOT NULL,\n" +
+                "  \"zip\" varchar(255) NOT NULL,\n" +
+                "  \"type\" enum('SHIPPING','BILLING','LIVING') NOT NULL,\n" +
+                "  PRIMARY KEY (\"id\"),\n" +
+                "  KEY \"address_customer\" (\"customer_id\"),\n" +
+                "  CONSTRAINT \"addresses_ibfk_1\" FOREIGN KEY (\"customer_id\") REFERENCES \"customers\" (\"id\")\n" +
+                ") ENGINE=InnoDB AUTO_INCREMENT=17 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci";
+        parser.parse(ddl, tables);
+        assertThat(parser.getParsingExceptionsFromWalker()).isEmpty();
+
+        Table table = tables.forTable(null, null, "addresses");
+        assertThat(table).isNotNull();
+        assertThat(table.columns()).hasSize(7);
+        assertThat(table.columnWithName("type")).isNotNull();
+        assertThat(table.columnWithName("type").typeName()).isEqualTo("ENUM");
+    }
+
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void shouldParseCreateTableWithDefaultModeDoubleQuotedStrings() {
+        String ddl = "CREATE TABLE t (\n" +
+                "  col1 ENUM(\"a\", \"b\", \"c\"),\n" +
+                "  col2 VARCHAR(10) DEFAULT \"test\"\n" +
+                ")";
+        parser.parse(ddl, tables);
+        assertThat(parser.getParsingExceptionsFromWalker()).isEmpty();
+
+        Table table = tables.forTable(null, null, "t");
+        assertThat(table).isNotNull();
+        assertThat(table.columns()).hasSize(2);
+    }
+
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void shouldParseSchemaQualifiedAlterTableWithAnsiQuotesMode() {
+        parser.parse("SET sql_mode='ANSI_QUOTES'", tables);
+        parser.parse("CREATE TABLE \"customers\" (\n" +
+                "  \"id\" int NOT NULL AUTO_INCREMENT,\n" +
+                "  PRIMARY KEY (\"id\")\n" +
+                ") ENGINE=InnoDB", tables);
+        parser.parse("ALTER TABLE inventory.\"customers\" ADD COLUMN \"middle_name\" varchar(255) NULL", tables);
+        assertThat(parser.getParsingExceptionsFromWalker()).isEmpty();
+    }
 }

--- a/debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/antlr/DdlNormalizerTest.java
+++ b/debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/antlr/DdlNormalizerTest.java
@@ -335,4 +335,63 @@ public class DdlNormalizerTest {
         String expected = "CREATE TABLE `RANK` (c VARCHAR(20) DEFAULT 'FROM RANK', `RANK` INT COMMENT 'RANK: n')";
         assertThat(DdlNormalizer.normalize(input)).isEqualTo(expected);
     }
+
+    @DisplayName("Given ANSI_QUOTES mode and double-quoted identifiers When normalize Then identifiers are preserved")
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void testAnsiQuotesModePreservesDoubleQuotedIdentifiers() {
+        String input = "CREATE TABLE \"customers\" (\"id\" int NOT NULL AUTO_INCREMENT, PRIMARY KEY (\"id\"))";
+        assertThat(DdlNormalizer.normalize(input, true)).isEqualTo(input);
+    }
+
+    @DisplayName("Given ANSI_QUOTES mode and full CREATE TABLE with double-quoted identifiers When normalize Then identifiers are preserved")
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void testAnsiQuotesModeFullCreateTable() {
+        String input = "CREATE TABLE \"addresses\" (\n" +
+                "  \"id\" int NOT NULL AUTO_INCREMENT,\n" +
+                "  \"customer_id\" int NOT NULL,\n" +
+                "  \"street\" varchar(255) NOT NULL,\n" +
+                "  \"city\" varchar(255) NOT NULL,\n" +
+                "  \"type\" enum('SHIPPING','BILLING','LIVING') NOT NULL,\n" +
+                "  PRIMARY KEY (\"id\"),\n" +
+                "  KEY \"address_customer\" (\"customer_id\"),\n" +
+                "  CONSTRAINT \"addresses_ibfk_1\" FOREIGN KEY (\"customer_id\") REFERENCES \"customers\" (\"id\")\n" +
+                ") ENGINE=InnoDB AUTO_INCREMENT=17 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci";
+        assertThat(DdlNormalizer.normalize(input, true)).isEqualTo(input);
+    }
+
+    @DisplayName("Given ANSI_QUOTES mode and ALTER TABLE with double-quoted identifiers When normalize Then identifiers are preserved")
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void testAnsiQuotesModeAlterTable() {
+        String input = "ALTER TABLE inventory.\"customers\" ADD COLUMN \"middle_name\" varchar(255) NULL";
+        assertThat(DdlNormalizer.normalize(input, true)).isEqualTo(input);
+    }
+
+    @DisplayName("Given ANSI_QUOTES mode with single-quoted strings When normalize Then strings are preserved")
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void testAnsiQuotesModeSingleQuotedStrings() {
+        String input = "CREATE TABLE \"t\" (\"col\" VARCHAR(10) DEFAULT 'hello')";
+        assertThat(DdlNormalizer.normalize(input, true)).isEqualTo(input);
+    }
+
+    @DisplayName("Given default mode (not ANSI_QUOTES) When normalize Then double-quoted strings are still converted")
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void testDefaultModeStillConvertsDoubleQuotes() {
+        String input = "CREATE TABLE t (col ENUM(\"a\", \"b\", \"c\"))";
+        String expected = "CREATE TABLE t (col ENUM('a', 'b', 'c'))";
+        assertThat(DdlNormalizer.normalize(input, false)).isEqualTo(expected);
+    }
+
+    @DisplayName("Given ANSI_QUOTES mode with reserved keywords When normalize Then keywords still get backticks")
+    @Test
+    @FixFor("debezium/dbz#2291")
+    public void testAnsiQuotesModeReservedKeywordsStillBackticked() {
+        String input = "CREATE TABLE RANK (\"id\" INT, \"name\" VARCHAR(50))";
+        String expected = "CREATE TABLE `RANK` (\"id\" INT, \"name\" VARCHAR(50))";
+        assertThat(DdlNormalizer.normalize(input, true)).isEqualTo(expected);
+    }
 }

